### DISCUSSION TITLE
R6 within S3 confusion

### DISCUSSION THREAD

**Discussion by dheimgartner** [davidgohel/officer#413] [Ideas]
Hi,

Thanks so much for this great package!

I would like to elaborate on one point that I find a little confusing and its consequence:

The confusion arises because of the following implementation:
- The main object generated with `officer::read_docx()` is of class `S3`
- However, within this `S3` object lives a field that is `R6` (namely the `x$doc_obj`)

Many of your functions directly act on this field (e.g. the `officer::body_replace_all_text` which thus directly modify the main object (reference semantics).

Here a simple example where `template` points to some local `.docx` file with the text **replace_this** appearing somewhere:

```{R}
test <- officer::read_docx(template)
officer::body_replace_all_text(test, "replace_this", "with_that")
officer::docx_summary(test)
```

I am not that knowledgeable in OO but I don't expect `S3` to have reference semantics. Maybe I am wrong in this point and one should expect this behaviour. Nevertheless, one thing that I find unlucky is that there is no obvious way to **not** change the original document. I.e.

```{R}
test <- officer::read_docx(template)
test_copy <- test
officer::body_replace_all_text(test, "replace_this", "with_that")
officer::docx_summary(test_copy)
```

changes `replace_this` even in `test_copy`.

(My idea was to iterate over a `.docx` letter template and substitute addresses, names, etc.)

One solution would be to change the main object (in the above example the `rdocx`) to an `R6`. Then one could use the `x$clone()` to initialize a clone and modify the clone without sideeffects...

Would be happy to hear your thoughts!

Best,
Daniel

**Comment by davidgohel** [comment]
I agree this is not convenient. 

This is something that could/should be improved. But I think everything should ideally use the `S3` formalism or even simple R functions. 

But for now, it's not really possible to clone/copy an 'officer' object. 

Feel free to propose a PR to clone an object, but I think it's not a clear part in the code (undocumented R6), and it could be difficult :/ 

I can add a TODO for that feature, but I can't promise I will dedicate time to this soon (sorry again...).

**Reply by dheimgartner** [reply]
Hi David,

Thanks for your response!

I was mostly interested in your opinion and understand that it is not urgent on the task list.

I agree, S3 would be cleanest (and then one could simply copy the object). As for me, I found a work around which does not require copying an officer object.

Best
Daniel
